### PR TITLE
fix(compiler): do not throw error if localExpr is instanceof ReadVarExpr

### LINKED ISSUE #32761: Changing value of custom directive in template causes prod build failing.
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/compiler

### Is this a regression?

 No

### Description

 If there is custom directive in template and you assign new value to it in template prod build fails.

## 🔬 Minimal Reproduction

Repository with example: https://github.com/mohaxspb/ng-var-directive-issue

To get error just run `ng build --prod`

## 🔥 Exception or Error
<pre><code>
ERROR in Cannot assign to a reference or variable!
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 8.3.4
Node: 10.15.3
OS: win32 x64
Angular: 9.0.0-next.6+80.sha-df1c456.with-local-changes
... animations, bazel, common, compiler, compiler-cli, core
... elements, forms, language-service, localize
... platform-browser, platform-browser-dynamic, platform-server
... platform-webworker, platform-webworker-dynamic, router
... service-worker, upgrade

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.4
@angular-devkit/build-angular     0.803.4
@angular-devkit/build-optimizer   0.803.4
@angular-devkit/build-webpack     0.803.4
@angular-devkit/core              8.3.4
@angular-devkit/schematics        8.3.4
@angular/cli                      8.3.4
@ngtools/webpack                  8.3.4
@schematics/angular               8.3.4
@schematics/update                0.803.4
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2
</code></pre>

**Anything else relevant?**

I'm in progress with PR with fix for it

### PR DESCRIPTION
**Pull request by mohaxspb**
Fixes #32761

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #32761 

## What is the new behavior?

`ng build --prod` doesn't fails.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

There is also one more PR with adding details to error message in this case: #32760

### CODE REVIEW

**Review by Splaktar — CHANGES_REQUESTED** [review]
There is a test to specifically verify that an exception is thrown here. I don't fully understand why `ReadVarExpr` is not supported here, but it seems to be by design? Are there any comments in the code that help to explain why this needs to throw?

```sh
HeadlessChrome 74.0.3729 (Linux 0.0.0) ChangeDetection expressions event expressions should throw when trying to assign to a local FAILED
	Error: Expected function to throw an Error.
	    at <Jasmine>
	    at UserContext.eval (packages/core/test/linker/change_detection_integration_spec.ts:667:16 <- angular/packages/core/test/linker/change_detection_integration_spec.js:517:28)
	    at UserContext.eval (node_modules/zone.js/dist/zone-testing.js:1442:34)
	    at ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:396:30)
```

If you want to propose changing this behavior, you'll want to change the test as well. 

We'll probably also need some more details in the PR description to explain why this is needed and what not allowing it will impact.

**Comment by AndrewKushnir** [comment]
@mohaxspb could you please also some test cases that were failing before and are now fixed? Thank you.

**Comment by mohaxspb** [comment]
@Splaktar , @AndrewKushnir , there is a story how I found this bug, may be it will be helpful to understand if we need this PR or not.

I want to be able to declare variable in template. I.e. `<div *myVar="someComponentVar as myTemplateVar">`. To be able to do it I found this code on [stackoverflow](https://stackoverflow.com/a/43172992/3212712)

```
@Directive({
  selector: '[myVar]',
})
export class VarDirective {
  @Input()
  set myVar(context: any) {
    this.context.$implicit = this.context.myVar = context;
    this.updateView();
  }

  context: any = {};

  constructor(private vcRef: ViewContainerRef, private templateRef: TemplateRef<any>) {
  }

  updateView() {
    this.vcRef.clear();
    this.vcRef.createEmbeddedView(this.templateRef, this.context);
  }
}
```

Then, I want to change variable value on some event. So I write this code: `<button (click)="myTemplateVar=someValue"></button>`

And it works in dev build. But, if I run `ng build --prod` I get error `ERROR in Cannot assign to a reference or variable!`. So I create question on [stackoverflow](https://stackoverflow.com/questions/57978475/angular-prod-build-error-in-cannot-assign-to-a-reference-or-variable/57979238) about it, as I can't find error cause with so poor error message. In answer I found a place in compiler, which throws error. I tried to fix it by adding one more condition and then everything is working as expected.

So I create 2 PR with changed error message to help developers to found error reason and this with fix. I don't really not if this PR can break something... So, if you think, that it's normal behavior - let's close this PR) But, I'm sure, that in case of closing it, we should merge #32760 with correct error message which can help in finding issue.

If you think, that this PR will not break something I could just remove test, which checks, that error is throwing.

**Comment by Splaktar** [comment]
I think that it will be easier for the team to evaluate if you add a new test that fails prior to this change, but passes with this change. It would likely include some code similar to your example above.

**Comment by mohaxspb** [comment]
> I think that it will be easier for the team to evaluate if you add a new test that fails prior to this change, but passes with this change. It would likely include some code similar to your example above.

OK, I'll try to do it)

File: packages/compiler/src/compiler_util/expression_converter.ts [hunk 1]
```diff
@@ -535,11 +535,11 @@ class _AstToIrVisitor implements cdAst.AstVisitor {
     const receiver: o.Expression = this._visit(ast.receiver, _Mode.Expression);
     const prevUsesImplicitReceiver = this.usesImplicitReceiver;
 
-    let varExpr: o.ReadPropExpr|null = null;
+    let varExpr: o.ReadPropExpr|o.ReadVarExpr|null = null;
     if (receiver === this._implicitReceiver) {
       const localExpr = this._getLocal(ast.name);
       if (localExpr) {
```
Comment (rbuzzibarrios) [inline packages/compiler/src/compiler_util/expression_converter.ts]: you may create a var with a name related with the condition you want to check inside ``if`` conditional.
for example:
```
const isPropExpOrVarExp = localExpr && (localExpr instanceof o.ReadPropExpr || localExpr instanceof o.ReadVarExpr);

if (isPropExpOrVarExp ) { 
   // Here your code.
}
```

Comment (dannyskoog) [reply packages/compiler/src/compiler_util/expression_converter.ts]: I would not recommend that. The reason is that we'll lose `type narrowing` within the `if`-clause.

And currently that's given to us due to the fact that we're utilizing `instanceof type guards` **within** the `if`-statement. If we were to encapsulate the results of these type guard expressions into a boolean variable (= **outside** the `if`statement), TypeScript won't be able to narrow down the type

**Comment by Splaktar** [comment]
Based on https://github.com/angular/angular/issues/32695#issuecomment-539730512, it looks like  it is correct to throw this error, it's only that the error message needs to be improved.

**Comment by mohaxspb** [comment]
> Based on [#32695 (comment)](https://github.com/angular/angular/issues/32695#issuecomment-539730512), it looks like it is correct to throw this error, it's only that the error message needs to be improved.

OK, I'll close it and try to finish PR with error message this week) 

Sorry for delay - I didn't have enough time for PRs.

**Pull request closed without merging**